File: Data Structures & Algorithms/two-integer-sum-ii/submission-2.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& numbers, int target) {
        int l = 0;
        int r = numbers.size() - 1;
        vector<int> res;
        while (r > l) {
            int sum = numbers[r] + numbers[l];
            if (sum == target) {
                res.push_back(l+1);
                res.push_back(r+1);
                return res;
            }
            else if (sum > target) {
                r--;
            }
            else {
                l++;
            }
        }
    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        cur = []
        cur_sum = 0
        length = len(candidates)

        def back(i):
            nonlocal cur_sum
            if cur_sum == target:
                res.append(cur.copy())
                return
            if i >= length or cur_sum > target:
                return
            
            cur.append(candidates[i])
            cur_sum += candidates[i]
            back(i+1)

            cur.pop()
            cur_sum -= candidates[i]

            while i + 1 < length and candidates[i] == candidates[i+1]:
                i+=1
            back(i+1)
        
        back(0)
        return res

